src/main/java/io/github/genomicdatainfrastructure/discovery/datasets/infrastructure/beacon/persistence/BeaconIndividualsRequestMapper.java

```
@@ -0,0 +1,125 @@
+    }
+
+    private BeaconIndividualsRequestQueryFilter buildBeaconFilter(DatasetSearchQueryFacet facet) {
+        return switch (facet.getType()) {
+            case DROPDOWN -> buildDropdownBeaconFacet(facet);
+            case FREE_TEXT -> buildFreeTextBeaconFacet(facet);
+            default -> null;
+        };
+    }
+
+    private BeaconIndividualsRequestQueryFilter buildDropdownBeaconFacet(
+            DatasetSearchQueryFacet facet) {
+        String value = ofNullable(facet.getValue())
+                .filter(v -> !v.isBlank())
+                .orElseThrow(() -> new BadRequestException(
+                        "Facet value must not be null or empty"));
+
+        return BeaconIndividualsRequestQueryFilter.builder()
+                .id(value)
+                .scope(SCOPE)
+                .build();
+    }
+
+    private BeaconIndividualsRequestQueryFilter buildFreeTextBeaconFacet(
+            DatasetSearchQueryFacet facet) {
+        String operator = ofNullable(facet.getOperator())
+                .map(it -> it.value())
+                .orElseThrow(() -> new BadRequestException(
+                        "Facet operator must not be null"));
+
+        String value = ofNullable(facet.getValue())
+                .filter(v -> !v.isBlank())
+                .orElseThrow(() -> new BadRequestException(
+                        "Facet value must not be null or empty"));
+
+        String key = ofNullable(facet.getKey())
+                .filter(k -> !k.isBlank())
+                .orElseThrow(() -> new BadRequestException("Facet key must not be null or empty"));
+
+        return BeaconIndividualsRequestQueryFilter
+                .builder()
+                .id(key)
+                .operator(operator)
+                .value(value)
+                .scope(SCOPE)
+                .build();
+    }
+
+    private Map<String, String> buildBeaconRequestParameters(DatasetSearchQueryFacet facet) {
+        var entries = ofNullable(facet
```

question: I beleive we could throw an exception if entries is empty. Wdyt?

---

Oh, actually you are doing that. Then you are missing only the check if the value is empty, it should not be.